### PR TITLE
Fix `Explain Analyze` crash

### PR DESCRIPTION
**Pull request by weinan003**
We have a kind of query that its `InitPlan` execute in `ExecutorStart`.
If `Explain Analyze` this kind of query, as well as, some memory or disk
information requires to collect in main plan, the QD will crash.
Since `queryDesc->showstatctx->stats_gathered` is assigned to `true` 
in `ExecSetParamPlan` function before `ExecutorRun`, So we only gather
`initplan` metrics and left other slices information over. Segment fault when
some execution node print out metrics message.

To fix this problem, variable `stats_gathered` only be assigned after slice 0
metrics information collection.
 
reproduce process:
```
create table t(a int);
explain analyze select a
from (
	select a
	from t
	where a > (
		select avg(a)
		from t
	)
) as foo
order by a
limit 1
```

fix issue:#6951

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/commands/explain_gp.c [hunk 1]
```diff
@@ -1948,7 +1953,7 @@ cdbexplain_showExecStatsEnd(struct PlannedStmt *stmt,
 			long mem_wanted;
 
 			mem_wanted = (long) PolicyAutoStatementMemForNoSpillKB(stmt,
-							(uint64) showstatctx->workmemwanted_max / 1024L);
+							(uint64) (showstatctx->workmemwanted_max + 1024) / 1024L);
```
Comment (danielgustafsson) [inline src/backend/commands/explain_gp.c]: This doesn't seem related to fixing the crash?

File: src/backend/commands/explain_gp.c [hunk 2]
```diff
@@ -772,8 +772,13 @@ cdbexplain_recvExecStats(struct PlanState *planstate,
...
 
-	/* Signal that we've gathered all the statistics */
-	showstatctx->stats_gathered = true;
+	/* Signal that we've gathered all the statistics
+	 * Since QD may executes Initplan before main plan,
+	 * main plan information skip to gather.
+	 * Thus, only set this variable on slice 0
+	 * */
+	if(sliceIndex == 0)
+		showstatctx->stats_gathered = true;
```
Comment (danielgustafsson) [inline src/backend/commands/explain_gp.c]: This seems reasonable. Small nitpick: there should be a space between `if` and the parenthesis like `if (sliceIndex == 0)`.

File: src/backend/commands/explain_gp.c [hunk 3]
```diff
@@ -772,8 +772,13 @@ cdbexplain_recvExecStats(struct PlanState *planstate,
 	/* Transfer worker counts to SliceSummary. */
 	showstatctx->slices[sliceIndex].dispatchSummary = ds;
 
-	/* Signal that we've gathered all the statistics */
-	showstatctx->stats_gathered = true;
+	/* Signal that we've gathered all the statistics
+	 * Since QD may executes Initplan before main plan,
+	 * main plan information skip to gather.
+	 * Thus, only set this variable on slice 0
+	 * */
```
Comment (danielgustafsson) [inline src/backend/commands/explain_gp.c]: I'm not sure I fully understand this comment, can you try to reword and extend it a little?

File: src/backend/commands/explain_gp.c (hunk 1 above)
Comment (weinan003) [reply src/backend/commands/explain_gp.c]: showstatctx->workmemwanted_max  some time will lower than 1024 if the data set is significantly small. An `Assert` failed in `PolicyAutoStatementMemForNoSpillKB`, after the crash fix, so I give `1K` at least.

Comment (danielgustafsson) [reply src/backend/commands/explain_gp.c]: I'm not a fan of this, skewing every calculation regardless of value (if ever so slightly) just to avoid a cornercase Assert seems like a hack. Since we are using a full uint64, there isn't really any need to insist in using kB though, I took a stab at a fix in #8288 which instead calculates in bytes to avoid this. What do you think about that?

Comment (weinan003) [reply src/backend/commands/explain_gp.c]: Yes, that is really much better! :)

Comment (danielgustafsson) [reply src/backend/commands/explain_gp.c]: Do you have a testcase handy for hitting this? I was doing very synthetic things (hacks), an actual repro would be nice.

Comment (weinan003) [reply src/backend/commands/explain_gp.c]:
```sql
create table t(a int); 
insert into t select * from generate_series(1,10); 
explain analyze select a
from (
	select a
	from t
	where a > (
		select avg(a)
		from t
	)
) as foo
order by a
limit 1;
```

Comment (danielgustafsson) [reply src/backend/commands/explain_gp.c]: doh, I didn't realize it was this very PR that was the repro, I was under the impression that you had stumbled across it when looking at this.

Comment (weinan003) [reply src/backend/commands/explain_gp.c]: lol

Comment (weinan003) [reply src/backend/commands/explain_gp.c]: It is very tricky, either no insert or `insert into... values ...` both can not trigger this problem. only use `generate_series` can find out

Comment (danielgustafsson) [reply src/backend/commands/explain_gp.c]: I have pushed #8288, if you rebase this PR on master without this line it should be good to go.
